File: backend/app/hooks/custom/example_hooks.py
"""
Example custom hooks.
This file demonstrates how to create custom hooks.
Remove or rename this file when creating your own hooks.
"""
from app.hooks import register_hook, HookType


@register_hook("product.before_create", HookType.BEFORE, priority=50)
def example_product_validation(product_data: dict, context: dict) -> dict:
    """
    Example: Validate product before creation.
    
    This hook runs before a product is created and can modify
    the product data or raise an exception to prevent creation.
    """
    # Example: Ensure product code has prefix
    if "code" in product_data and product_data["code"]:
        if not product_data["code"].startswith("PROD-"):
            product_data["code"] = f"PROD-{product_data['code']}"
    
    return product_data


@register_hook("product.after_create", HookType.AFTER)
def example_product_notification(product, context: dict) -> None:
    """
    Example: Send notification after product creation.
    
    This hook runs after a product is created and can perform
    side effects like sending notifications or logging.
    """
    # Example: Log product creation
    print(f"Product created: {product.name} (ID: {product.id})")
    # In real implementation, you might send email, webhook, etc.


@register_hook("order.filter_total", HookType.FILTER)
def example_order_discount(order, context: dict) -> float:
    """
    Example: Apply custom discount to order total.
    
    This hook can modify the order total calculation.
    """
    # Example: Apply 5% discount for orders over $100
    if order.subtotal > 100:
        return order.subtotal * 0.95
    return order.subtotal


@register_hook("payment.before_process", HookType.BEFORE, priority=10)
def example_payment_validation(payment_data: dict, context: dict) -> dict:
    """
    Example: Validate payment before processing.
    
    This hook runs before payment is processed and can modify
    payment data or raise an exception to prevent processing.
    """
    # Example: Ensure minimum payment amount
    if payment_data.get("amount", 0) < 0.01:
        raise ValueError("Payment amount must be at least $0.01")
    
    return payment_data

File: frontend/console/src/hooks-system/custom/example-hooks.ts
/**
 * Example custom hooks.
 * This file demonstrates how to create custom hooks.
 * Remove or rename this file when creating your own hooks.
 */

import { registerHook, HookType } from '../index';
import { ProductFormData, OrderFormData } from '../types';

// Example: Product form validation
registerHook('product.beforeSubmit', HookType.BEFORE, (data: ProductFormData) => {
  // Ensure product code has prefix
  if (data.code && !data.code.startsWith('PROD-')) {
    data.code = `PROD-${data.code}`;
  }
  
  return data;
}, 50);

// Example: Product display customization
registerHook('product.afterLoad', HookType.AFTER, (product: any) => {
  // Add custom display name
  return {
    ...product,
    displayName: `${product.name} (${product.code || 'N/A'})`
  };
});

// Example: Filter product list
registerHook('product.filterList', HookType.FILTER, (products: any[]) => {
  // Filter and sort products
  return products
    .filter(p => p.isActive)
    .sort((a, b) => a.name.localeCompare(b.name));
});

// Example: Additional form fields
registerHook('product.formFields', HookType.COMPONENT, () => {
  // Return additional form field component
  // In real implementation, import and return your custom component
  return null; // Placeholder
});

// Example: Order total calculation
registerHook('order.calculateTotal', HookType.FILTER, (order: OrderFormData) => {
  // Apply custom discount for orders over $100
  const subtotal = order.items.reduce((sum, item) => sum + (item.quantity * item.unitPrice), 0);
  if (subtotal > 100) {
    // Apply 5% discount
    return subtotal * 0.95;
  }
  return subtotal;
});

